### PR TITLE
Add remove columns panel

### LINKED ISSUE #33841: Introduce Add/Remove panel
This issue really encompasses the majority of the Add / Remove column Epic

### PR DESCRIPTION
**Pull request by npfitz**
Closes https://github.com/metabase/metabase/issues/33841

### Description
This PR adds a new panel to control adding and removing table columns from a query, separating the idea of add/remove from show/hide in viz settings. The new panel is only available in structured queries (and should not be displayed for dashcards), and should allow you to implicitly add fields from related tables, search available columns, and perform bulk add / remove column operations per table.

### How to verify
Create a structured question and try adding / removing columns from a query. A few points:
- Showing / Hiding should not re-run queries
- Try using saved questions as table sources
- You should now be able to hide columns from the table header

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: frontend/src/metabase/visualizations/components/settings/ColumnItem/ColumnItem.tsx
Comment (npfitz) [inline frontend/src/metabase/visualizations/components/settings/ColumnItem/ColumnItem.tsx:50]: One of the larger sins introduced in this PR. I know we don't typically use these, but it made targeting disabled columns via CSS so much easier. In the future, it would be good to pass elements to the sorting components, rather than just objects used to render these column items.

File: src/metabase/lib/metadata/calculation.cljc
Comment (npfitz) [inline src/metabase/lib/metadata/calculation.cljc:360]: @bshepherdson I'll need your blessing for these added props

Comment (bshepherdson) [reply src/metabase/lib/metadata/calculation.cljc:360]: LGTM

**Comment by alxnddr** [comment]
Hey @npfitz, I found a couple of cases that are not covered with specs, can you please add these?
1) Inability to reenable custom column anywhere except the table column settings
https://github.com/metabase/metabase/assets/14301985/f2ad26c8-8336-42df-b70e-54fc93dcd777

2) Implicitly joinable tables incorrectly show columns selected. Both models are just the Orders table — one is native, another is structured. To repro this bug, you need to start the question with the native one and join the structured

https://github.com/metabase/metabase/assets/14301985/ca02ef7b-4c38-4c38-96f3-687d86a6fb86

3) Minor cosmetic issue: sometimes it flushes wrong selection before applying and showing the right one
https://github.com/metabase/metabase/assets/14301985/6ca34e51-d865-4a22-a76a-78a740bd8e74

**Comment by iethree** [comment]
Took this for a quick spin, very slick!

Found a couple things

It looks like all column selection gets disabled when there's an aggregation in the query:

aggBreakColSelect

Also, for aggregated queries, I can't change any viz settings from column headers at all

colVizSetDisabled

Also, for tables with lots of columns (my pokemon card table has 43), hiding columns is really slow

lag hide col

profiler tells me the UI locks up for 1.7s

Screen Shot 2023-10-17 at 7 50 21 PM

**Comment by npfitz** [comment]
@alxnddr Looking into your feedback now!
1. It looks like this has been a thing for a while. I pulled down 47 and tested and it did the same thing. Weirdly enough, if you hide one of the other columns (like Count), it hides it on all stages of the preview. Super weird
2. It looks like in the past, we wouldn't have shown implicitly joinable columns for a question like this. Still looking into why, but I think the main thing here is that We have more work to do with matching up a column to a setting
3. Not totally sure what's going on there :/

**Comment by alxnddr** [comment]
I noticed another case, when users change settings and then the query changes it leads to an error inside the sidebar:

https://github.com/metabase/metabase/assets/14301985/dad4117a-a5b0-44cd-8afc-6b4e10d0e918

**Review by iethree — CHANGES_REQUESTED** [review]
**Broken Viz Settings in Column Headers**

I realize the fact that the secondary-table view that doesn't observe viz settings is an inherited problem, but it feels like adding more viz settings buttons to that view is making more bugs, and making them more obvious.

I don't feel good about merging this unless (a) we hide viz settings in column headers when display is set to something other than table, or (b) we make that view actually obey table viz settings.

**Performance**

This is also an inherited problem, but testing with ~40 columns, I'm seeing it take 10+ seconds to hide a column from the table header. This locks up the whole ui. I realize this was the existing behavior for hiding columns, but surfacing this performance problem in a "chill mode" interface like column headers feels really bad.

Screen Shot 2023-10-25 at 2 48 28 PM


Screen Shot 2023-10-25 at 2 43 49 PM

But really, these are product concerns, and @luizarakaki should weigh in on how to deal with them.

File: frontend/src/metabase/visualizations/components/settings/ChartSettingAddRemoveColumns/ChartSettingAddRemoveColumns.unit.spec.tsx
Comment (alxnddr) [inline frontend/src/metabase/visualizations/components/settings/ChartSettingAddRemoveColumns/ChartSettingAddRemoveColumns.unit.spec.tsx:76]: Can we add a bit more specs on different corner cases here? Historically, this functionality was  a bit fragile so having more specs would be a huge quality improvement

File: frontend/src/metabase/visualizations/components/ChartSettings.jsx
Comment (alxnddr) [inline frontend/src/metabase/visualizations/components/ChartSettings.jsx:100]: We are introducing the concept of `widgetOverride` with `writeSettingId` to the shared settings code only for one specific place. Would it be possible to still have the original settings definition but its widget would be a component that switches between two views?

File: frontend/src/metabase/visualizations/components/settings/ChartSettingAddRemoveColumns/ChartSettingAddRemoveColumns.tsx
Comment (alxnddr) [inline frontend/src/metabase/visualizations/components/settings/ChartSettingAddRemoveColumns/ChartSettingAddRemoveColumns.tsx:163]: Missing translation

File: e2e/test/scenarios/question/settings.cy.spec.js
Comment (alxnddr) [inline e2e/test/scenarios/question/settings.cy.spec.js:177]: Can we remove this line?

File: e2e/test/scenarios/visualizations/table-column-settings.cy.spec.js
Comment (alxnddr) [inline e2e/test/scenarios/visualizations/table-column-settings.cy.spec.js:280]: Using underscore would be a bit more consistent

**Review by alxnddr — COMMENTED** [review]
Huge UX improvement! I left a few comments/question codewise

File: frontend/src/metabase/visualizations/components/settings/ChartSettingAddRemoveColumns/ChartSettingAddRemoveColumns.unit.spec.tsx
Comment (npfitz) [reply frontend/src/metabase/visualizations/components/settings/ChartSettingAddRemoveColumns/ChartSettingAddRemoveColumns.unit.spec.tsx:76]: I can try to add more here, but many corner cases on things like question structure are handled in `table-column-settings.cy.spec.js`

**Comment by iethree** [comment]
@npfitz what do you think about https://github.com/metabase/metabase/pull/35114 ?
It addresses one of my concerns about the broken hide button.

**Comment by npfitz** [comment]
> I noticed another case, when users change settings and then the query changes it leads to an error inside the sidebar:

@alxnddr This is a bit awkward, but this is happening because now clicking the eye only hides the column, so we don't remove it from `table.columns`, and we also don't remove them when we remove the fields from the query via the summarize bar (This is actually related to https://github.com/metabase/metabase/pull/34976). I confirmed that stacking the logic from that PR onto this will solve the issue, but I don't know that we want to bloat this PR more than it already is. I think merging one, then the other, is the right thing to do here

**Comment by alxnddr** [comment]
@npfitz awesome you already got a fix for this, let's merge them separately then

File: frontend/src/metabase/visualizations/components/settings/ChartSettingTableColumns/utils.ts
Comment (alxnddr) [inline frontend/src/metabase/visualizations/components/settings/ChartSettingTableColumns/utils.ts:132]: Forgotten comment

**Review by alxnddr — APPROVED** [review]

**Review by iethree — APPROVED** [review]

**Pull request merged**
